### PR TITLE
More sane buffer conversion with better memory usage

### PR DESCRIPTION
**Pull request by rzezeski**
az608 bz1065

Convert the awkward buffer->segment process to a more OTP compliant
simple_one_for_one supervisor.  To keep the previous behavior of only one
buffer at a time per mi_server I made use of a queue in the mi_server to
keep track of pending buffer conversions.

This reduces memory usage as it now completely avoids long-lived buffer
conversion processes.

### CODE REVIEW

**Comment by rzezeski** [comment]
To test this I used a single node (`make rel`) and indexed 5k text/plain docs ~5.5KB in size under a KV bucket with search hook installed.

Below are the results for `erlang:memory/0` before/after indexing **before** this patch.

```
total:              18MB -> 228MB
processes:          5MB -> 70MB
processes_u         sed: 5MB -> 70MB
system:             14MB -> 159MB
atom:               1MB -> 1MB
binary:             0MB -> 17MB
code:               8MB -> 9MB
ets:                1MB -> 127MB
```

Below is the memory dominator list after indexing.  That is, this list represents the processes using the most memory in the system and what `mod:fun` they were started with.  Notice that the list is dominated by `mi_buffer_converter`.

```
(riak@127.0.0.1)33> io:format("~p~n", [Dom3]).
[{4114768,<0.678.0>,{mi_buffer_converter,init,1}},
 {4114768,<0.607.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.694.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.690.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.662.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.627.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.562.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.507.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.497.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.487.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.467.0>,{mi_buffer_converter,init,1}},
 {2543424,<0.427.0>,{mi_buffer_converter,init,1}},
 {1572280,<0.698.0>,{mi_buffer_converter,init,1}},
 {1572280,<0.623.0>,{mi_buffer_converter,init,1}},
 {1572280,<0.592.0>,{mi_buffer_converter,init,1}},
 {1572280,<0.567.0>,{mi_buffer_converter,init,1}},
 {1572280,<0.537.0>,{mi_buffer_converter,init,1}},
 {1572280,<0.482.0>,{mi_buffer_converter,init,1}},
 {972080,<0.682.0>,{mi_buffer_converter,init,1}},
 {972080,<0.492.0>,{mi_buffer_converter,init,1}},
 {972080,<0.457.0>,{mi_buffer_converter,init,1}},
 {972080,<0.422.0>,{mi_buffer_converter,init,1}},
 {601136,<0.658.0>,{mi_buffer_converter,init,1}},
 {601136,<0.615.0>,{mi_buffer_converter,init,1}},
 {601136,<0.611.0>,{mi_buffer_converter,init,1}},
 {601136,<0.602.0>,{mi_buffer_converter,init,1}},
 {601136,<0.597.0>,{mi_buffer_converter,init,1}},
 {601136,<0.572.0>,{mi_buffer_converter,init,1}},
 {601136,<0.532.0>,{mi_buffer_converter,init,1}},
 {601136,<0.442.0>,{mi_buffer_converter,init,1}},
 {601136,<0.417.0>,{mi_buffer_converter,init,1}},
 {601072,<0.68.0>,undefined},
 {371880,<0.674.0>,{mi_buffer_converter,init,1}},
 {371880,<0.670.0>,{mi_buffer_converter,init,1}},
 {371880,<0.666.0>,{mi_buffer_converter,init,1}},
 {371880,<0.619.0>,{mi_buffer_converter,init,1}},
 {371880,<0.587.0>,{mi_buffer_converter,init,1}},
 {371880,<0.582.0>,{mi_buffer_converter,init,1}},
 {371880,<0.577.0>,{mi_buffer_converter,init,1}},
 {371880,<0.517.0>,{mi_buffer_converter,init,1}},
 {371880,<0.512.0>,{mi_buffer_converter,init,1}},
 {371880,<0.452.0>,{mi_buffer_converter,init,1}},
 {371880,<0.447.0>,{mi_buffer_converter,init,1}},
 {230192,<0.654.0>,{mi_buffer_converter,init,1}},
 {230192,<0.552.0>,{mi_buffer_converter,init,1}},
 {230192,<0.502.0>,{mi_buffer_converter,init,1}},
 {230192,<0.462.0>,{mi_buffer_converter,init,1}},
```

Next is the delta list before/after indexing **after** this patch was put in place.

```
total:              17MB -> 148MB
processed:          5MB -> 15MB
processes_used:     5MB -> 15MB
system:             13MB -> 133MB
atom:               1MB -> 1MB
code:               8MB -> 9MB
ets:                1MB -> 111MB
```

Below is the dominator tree after indexing.  Notice there is no longer a `mi_buffer_converter` anywhere in site.

```
(riak@127.0.0.1)23> io:format("~p~n", [Dom2]).
[{601072,<0.68.0>,undefined},
 {147232,<0.289.0>,{riak_core_vnode_master,init,1}},
 {142776,<0.513.0>,{riak_core_vnode,init,1}},
 {142776,<0.501.0>,{riak_core_vnode,init,1}},
 {142776,<0.471.0>,{riak_core_vnode,init,1}},
 {142776,<0.468.0>,{riak_core_vnode,init,1}},
 {142776,<0.462.0>,{riak_core_vnode,init,1}},
 {142776,<0.450.0>,{riak_core_vnode,init,1}},
 {142776,<0.444.0>,{riak_core_vnode,init,1}},
 {142776,<0.432.0>,{riak_core_vnode,init,1}},
 {142776,<0.429.0>,{riak_core_vnode,init,1}},
 {142776,<0.426.0>,{riak_core_vnode,init,1}},
 {142776,<0.423.0>,{riak_core_vnode,init,1}},
 {142776,<0.414.0>,{riak_core_vnode,init,1}},
 {142776,<0.408.0>,{riak_core_vnode,init,1}},
 {142776,<0.405.0>,{riak_core_vnode,init,1}},
 {142776,<0.396.0>,{riak_core_vnode,init,1}},
 {142776,<0.390.0>,{riak_core_vnode,init,1}},
 {142776,<0.388.0>,{riak_core_vnode,init,1}},
 {142776,<0.386.0>,{riak_core_vnode,init,1}},
 {142776,<0.384.0>,{riak_core_vnode,init,1}},
 {142776,<0.383.0>,{riak_core_vnode,init,1}},
 {142776,<0.382.0>,{riak_core_vnode,init,1}},
 {142776,<0.380.0>,{riak_core_vnode,init,1}},
 {142776,<0.379.0>,{riak_core_vnode,init,1}},
 {142776,<0.378.0>,{riak_core_vnode,init,1}},
 {142776,<0.376.0>,{riak_core_vnode,init,1}},
 {142776,<0.373.0>,{riak_core_vnode,init,1}},
 {142776,<0.372.0>,{riak_core_vnode,init,1}},
 {142776,<0.370.0>,{riak_core_vnode,init,1}},
 {142624,<0.295.0>,{supervisor,riak_kv_put_fsm_sup,1}},
 {142560,<0.3.0>,undefined},
 {142520,<0.25.0>,undefined},
```

File: src/mi_server.erl [hunk 1]
```diff
@@ -521,10 +530,21 @@ handle_info({'EXIT', CompactingPid, Reason},
...
-handle_info({'EXIT', ConverterSup, Reason},
-            #state{buffer_converter={ConverterSup, _}}=State) ->
-    %% if our converter's supervisor died, there's a problem: exit
-    {stop, {buffer_converter_death, Reason}, State};
+
+handle_info({'EXIT', Converter, Reason},
+            #state{root=Root,
+                   converter=Converter, to_convert=ToConvert}=State) ->
+    case Reason of
+        normal -> {noreply, State};
```
Comment (Vagabond) [inline src/mi_server.erl]: shutdown is also a valid reason to exit.

**Comment by rzezeski** [comment]
Changed converter to plain erlang but seeing these gems in log.

```
23:17:31.988 [warning] `convert` attempted to work with a nonexistent buffer, probably because drop was called [{ets,match_object,[426010,'_']},{ets,'tab2list',1},{mi_buffer,iterator,1},{mi_segment,from_buffer,2},{mi_buffer_converter,convert,4},{proc_lib,init_p_do_apply,3}]
23:17:31.989 [error] CRASH REPORT Process [] with 0 neighbours crashed with reason: buffer_dropped
23:17:31.990 [error] Supervisor mi_buffer_converter_sup had child at module undefined at <0.855.0> exit with reason


23:17:35.142 [error] CRASH REPORT Process [] with 0 neighbours crashed with reason: {badmatch,{error,{not_a_log_file,"/tmp/test.test-test/prop_api/segment.2.offsets"}}}
23:17:35.144 [error] Supervisor mi_buffer_converter_sup had child at module undefined at <0.2525.0> exit with reason {badmatch,{error,{not_a_log_file,"/tmp/test.test-test/prop_api/segment.1.offsets"}}} in context child_terminated
```

I figure the 1st has to do w/ the fact that I'm using plain Erlang/proc_lib?  The 2nd has me baffled as that path is for a segment but that error atom `not_a_log_file` is from the OTP `kernel` application.

**Comment by rzezeski** [comment]
Alright, things are looking better now.  No more inexplicable errors in the log file when running the tests.  Tag, your it, Andrew.

**Comment by Vagabond** [comment]
+1 from me on this, looks much better.

**Pull request merged**